Draft: entrypoint: Fix getting last index for tables

VladyslavTupikin opened the pull request:
Current implementation skips correct backend index for last object and use incorrect one. This is triggering problem in Web UI and CLI when we can't see correct data from the data model because wrong index saved in DB.

For solving this issue need to modify conditions: need to use the correct backend index and modify SQL query to the DB for saving data.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

MuteSpirit commented on src/ep_worker.c:
```diff
@@ -9750,6 +9750,7 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
     char *selectFromTempVal = NULL;
     BOOL subst_other_obj = FALSE;
     BOOL is_first_element = FALSE;
+    int  BackendIndex = -1;
```
Code style violation - naming of variables should be in `snake_style`, so i suggest to rename `BackendIndex -> backend_index`

MuteSpirit commented on src/ep_worker.c:
```diff
@@ -9823,6 +9824,11 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
     tmp_cnt = 0;
     for (i = 0; i < parsed_backend_str->subst_val_num; i++)
     {
+        if (parsed_backend_str->subst_val[i].mmx_subst_val.leaf_name, idx_params[idx_params_num - 1]) {
```
I don't understand the goal of expression like
```C
if (a, b) {
   // do something
}
```
Why not use
```C
if (idx_params[idx_params_num - 1]) {
```

nmerinov replied:
I believe it was `if (0 == strcmp(....))`

VladyslavTupikin replied:
fixed

MuteSpirit commented on src/ep_worker.c:
```diff
@@ -9848,18 +9854,39 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
+    {
+
+        strcat_safe(query, "IFNULL('", query_size);
+        strcat_safe(query, newrow->be_keys[BackendIndex], query_size);
```
What happen i we go inside this if branch when `BackendIndex` equal `-1`?

MuteSpirit reviewed (changes requested):
Commit comment does not describe made changes - i cannot understand that
Could you please add into almost commit comment the example wrong and good SQL query for your case?

MuteSpirit commented on src/ep_worker.c:
```diff
@@ -9848,18 +9854,39 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
-    strcat_safe(query, "]),0)+1 ", query_size);
+        strcat_safe(query, "[", query_size);
+        strcat_safe(query, idx_params[idx_params_num-1], query_size);
+        strcat_safe(query, "]),0)+1 ", query_size);
```
Is `+1` needed for case when we use inplicit `XIndex` field, for example, `BSSIndex` and return it from getall backend script?

nmerinov replied:
The whole logic if following:

When we add keys to the table we made it in the following manner (e.g. for Device.A.{i}.B.{i}.C.{i}., and getall call with Device.A.{i}.KeyA, Device.A.{i}.B.{i}.KeyB, KeyC answer):
```
INSERT INTO TableC_ValuesTbl 
  (AIndex, BIndex, CIndex, KeyC)
  SELECT (A.AIndex, B.BIndex, (MAX(C.CIndex)+1), KeyCValueFromGetallAnswer)
  FROM
    TableC_ValuesTbl as C
    LEFT JOIN TableB_ValuesTbl as B ON (C.BIndex = B.BIndex)
    LEFT JOIN TableA_ValuesTbl as A ON (C.AIndex = A.AIndex)
  WHERE
    B.KeyB = KeyBValueFromGetallAnswer
    AND A.KeyA = KeyAValueFromGetallAnswer
```

Currently we in situation when `KeyC` is actually `CIndex`. Idea of the changes is to use `KeyCValueFromGetallAnswer` for `CIndex` instead of `(MAX(C.CIndex)+1)` in select request.

nmerinov commented on src/ep_worker.c:
```diff
@@ -9848,18 +9854,39 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
     /* The last index is calculated in the object's table. */
-    if (idx_params_num > 1)
-        strcat_safe(query, "IFNULL(MAX(t0.", query_size);
+     if (BackendIndex == -1 || subst_other_obj || idx_params_num > 1)
```
What is it? I suggested to use this condition around FROM instead of hack with `MAX` to select something from empty table. It useless in this case.

VladyslavTupikin replied:
fixed

MuteSpirit commented on src/ep_worker.c:
```diff
@@ -9848,14 +9855,17 @@ static ep_stat_t w_form_query_getall_insert_new(worker_data_t *wd, obj_info_t *o
     /* The last index is calculated in the object's table. */
+
+    strcat_safe(query, "IFNULL('". query_size);
+    strcat_safe(query, newrow->be_keys[backend_index], query_size);
```
1) You should take into account that `backend_index` may be equal `-1` here
2) It's unclear from code what `newrow->be_keys[backend_index]` means
    I suggest to add explanation comment for that

Commit:
mmx-ep: Trim whitespaces in ep_worker.c

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
entrypoint: Fix getting last index for tables

Current implementation skips correct backend index
for last object and use incorrect one. This is
triggering problem in Web UI and CLI when we can't
see correct data from the data model because wrong
index saved in DB.

For solving this issue need to modify conditions:
need to use the correct backend index and modify
SQL query to the DB for saving data.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

VladyslavTupikin commented:
This workaround can't be used in main branch.
